### ISSUE TITLE
normalization ICE during codegen of trait objects with supertraits

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust#56229]
The following example ICEs 

```rust
impl Mirror for Even {
    type Other = Odd;
}

impl Mirror for Odd {
    type Other = Even;
}

trait Dyn<T: Mirror>: AsRef<<T as Mirror>::Other> { }

impl Dyn<Odd> for Even { }

impl AsRef<Even> for Even {
    fn as_ref(&self) -> &Even {
        self
    }
}

fn code<T: Mirror>(d: &dyn Dyn<T>) -> &T::Other {
    d.as_ref()
}

fn main() {
    println!("{:?}", code(&Even(22)));
}
```

yields

```
athena. rustc +nightly ~/tmp/as-ref-bug.rs
error: internal compiler error: src/librustc/traits/codegen/mod.rs:68: Encountered error `Unimplemented` selecting `Binder(<dyn Dyn<Odd> as std::convert::AsRef<Even>>)` during codegen

thread 'main' panicked at 'Box<Any>', src/librustc_errors/lib.rs:600:9
note: Run with `RUST_BACKTRACE=1` for a backtrace.
error: aborting due to previous error


note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.32.0-nightly (6acbb5b65 2018-11-25) running on x86_64-unknown-linux-gnu
```

(Same is true on stable.)

**Comment by arielb1** [comment]
That's because the ObjectCandidate code is not normalizing the upcast trait refs.

Should we wait for Chalk for the fix?

**Comment by kuviman** [comment]
Hitting this error, any updates?

**Comment by Centril** [comment]
Triage: no longer ICEs.

**Comment by JohnTitor** [comment]
It's still an ICE, note that we should build/run the snippet to check it ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=179337c2ff3cebb633e8c7720da1e46b)).

**Comment by JohnTitor** [comment]
Triage: It's no longer ICE with the latest nightly, I think it's fixed by #77720.
